File: 09.Greed贪心/leetcode_1288_removeCoveredIntervals.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    int removeCoveredIntervals(vector<vector<int>>& intervals) {
        sort(intervals.begin(), intervals.end(), [](auto&a, auto&b){
            return a[0]<b[0] || (a[0]==b[0] && a[1]>b[1]);
        });
        int res = intervals.size();
        int maxRight = 0;
        for (int i=0; i<intervals.size(); i++){
            int l = intervals[i][0], r = intervals[i][1];
            if (r <= maxRight){
                res--;
            }
            else{
                maxRight = r;
            }
        }
        return res;
    }
};

int main(){
    Solution s;
    vector<vector<int>> intervals = {{1,4},{3,6},{2,8}};
    cout << s.removeCoveredIntervals(intervals) << endl;
}

File: 09.Greed贪心/leetcode_1288_removeCoveredIntervals.py
from typing import List

class Solution:
    def removeCoveredIntervals(self, intervals: List[List[int]]) -> int:
        intervals.sort(key = lambda x:(x[0], -x[1]))
        res = len(intervals)
        maxRight = 0
        for l, r in intervals:
            if r <= maxRight:
                res -= 1
            else:
                maxRight = r
        return res

if __name__ == "__main__":
    s = Solution()
    intervals = [[1,4],[3,6],[2,8]]
    print(s.removeCoveredIntervals(intervals))
